src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -77,48 +112,110 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
 
         public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
         {
-            //
-            // We lower an interpolated string into an invocation of String.Format.  For example, we translate the expression
-            //
-            //     $"Jenny don\'t change your number { 8675309 }"
-            //
-            // into
-            //
-            //     String.Format("Jenny don\'t change your number {0}", new object[] { 8675309 })
-            //
-
             Debug.Assert(node.Type.SpecialType == SpecialType.System_String); // if target-converted, we should not get here.
-            BoundExpression format;
-            ArrayBuilder<BoundExpression> expressions;
-            MakeInterpolatedStringFormat(node, out format, out expressions);
-            if (expressions.Count == 0)
+
+            BoundExpression result;
+
+            if (CanLowerToStringConcatenation(node))
             {
-                // There are no fill-ins. Handle the escaping of {{ and }} and return the value.
-                Debug.Assert(!format.HasErrors && format.ConstantValue != null && format.ConstantValue.IsString);
-                var builder = PooledStringBuilder.GetInstance();
-                var formatText = format.ConstantValue.StringValue;
-                int formatLength = formatText.Length;
-                for (int i = 0; i < formatLength; i++)
+                // All fill-ins, if any, are strings, and none of them have alignment or format specifiers.
+                // We can lower to a more efficient string concatenation
+                // The normal pattern for lowering is to lower subtrees before the enclosing tree. However in this case
+                // we want to lower the entire concatenation so we get the optimizations done by that lowering (e.g. constant folding).
+
+                int length = node.Parts.Length;
+                if (length == 0)
                 {
-                    char c = formatText[i];
-                    builder.Builder.Append(c);
-                    if ((c == '{' || c == '}') && (i + 1) < formatLength && formatText[i + 1] == c)
+                    // $"" -> ""
+                    return _factory.StringLiteral("");
+                }
+
+                if (length == 1)
+                {
+                    var part = node.Parts[0];
+
+                    if (part is BoundStringInsert fillin)
+                    {
+                        // $"{part}"
+                        // this is the filled-in expression
+                        part = fillin.Value;
+
+                        if (part.ConstantValue != null)
+                        {
+                            if (part.ConstantValue.StringValue == null)
+                            {
+                                return _factory.StringLiteral("");
```

Comment:
>_factory.StringLiteral("") [](start = 39, length = 26)

It seems like `_factory.StringLiteral("")` always creates instances. @gafter Do you think it's worth it to use a static singleton here?

Reply:
No, this is fine. It is easier to debug if the bound nodes have the right syntax node associated with them, which the factory helps with.

---
In reply to: [190383830](https://github.com/dotnet/roslyn/pull/26612#discussion_r190383830) [](ancestors = 190383830)